### PR TITLE
low rank model trainer

### PR DESCRIPTION
**Pull request by deerzq**
Trainer for the low-rank model

@hectorgon

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    val pointwise = options.cache match {
+      case "memory" => raw.cache()
+      case _ : String => raw
+    }
+
+    val model = setupModel(options, pointwise)
+
+    modelIteration(sc, options, model, pointwise)
+
+    filterZeros(model)
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: I don't think you will get zero rows without L1 regularization

File: training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+  private def uniformRankLoss(k: Int, dim: Int): Float = {
+    // dim is the number of classes
+    // assume we always have dim > 1
+    k * 1.0f / (dim - 1.0f)
+  }
+
+  private def nonUniformRankLoss(k: Int, dim: Int): Float = {
+    var loss = 0.0f
+    for (i <- 1 to k) {
+      loss += 1.0f / i
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: For efficiency this can probably be computed once and passed as an array

File: training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        entry.setCount(kv._2.count.toInt)
+        dict.add(entry)
+      } else {
+        if (!featureWeights.containsKey(family)) {
+          featureWeights.put(family, new java.util.HashMap[java.lang.String, FloatVector]())
+        }
+        val familyMap = featureWeights.get(family)
+        if (!familyMap.containsKey(feature)) {
+          count = count + 1
+          familyMap.put(feature, FloatVector.getGaussianVector(embeddingSize))
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: Ideally this should be uniformly drawn from +/- 1/sqrt(embeddingSize)
http://stats.stackexchange.com/questions/47590/what-are-good-initial-weights-in-a-neural-network

**Comment by hectorgon** [comment]
LGTM with some minor comments regarding initialization and L1 regularization

**Comment by deerzq** [comment]
@hectorgon since we are not going to use sparse boost solver, perhaps we don't need to compute the real norm in GradientContainer? this can save some computation

**Comment by hectorgon** [comment]
Yes no need to use the real norm that was just for debugging anyway. One good check for convergence is to pick a random feature (or label) and print the top K labels close to that feature/label.
e.g. for twenty news alt.atheism should be near alt.religion. This is optional mostly for debugging. One could also supply a list of labels to compute nearest neighbors to for debugging for convergence at each iteration

File: training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala (hunk 3 above)
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: Thanks, updated the PR to use uniform sampling for initialization

**Comment by deerzq** [comment]
removed the feature norm computation part since it's not used in rprop solver

File: training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala (hunk 2 above)
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: added a `todo` for this task in the code

File: training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    }
+  }
+
+  private def uniformRankLoss(k: Int, dim: Int): Float = {
+    // dim is the number of classes
+    // assume we always have dim > 1
+    k * 1.0f / (dim - 1.0f)
+  }
+
+  private def nonUniformRankLoss(k: Int, dim: Int): Float = {
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: I think this is called "weighted approximate ranking loss" or WARP loss.

Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: this term is defined in the wsabie paper, I guess all three losses are 'WARP' loss.
